presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -442,42 +467,45 @@ private void loadPartition(HivePartitionMetadata partition)
             Map<Integer, HiveType> columnCoercions)
             throws IOException
     {
-        ImmutableList.Builder<HiveSplit> builder = ImmutableList.builder();
-
         boolean forceLocalScheduling = HiveSessionProperties.isForceLocalScheduling(session);
 
         if (splittable) {
-            for (BlockLocation blockLocation : blockLocations) {
-                // get the addresses for the block
-                List<HostAddress> addresses = toHostAddress(blockLocation.getHosts());
+            PeekingIterator<BlockLocation> blockLocationIterator = Iterators.peekingIterator(Arrays.stream(blockLocations).iterator());
 
-                long maxBytes = maxSplitSize.toBytes();
-                boolean creatingInitialSplits = false;
+            return new AbstractIterator<HiveSplit>() {
+                private long chunkOffset = 0;
 
-                if (remainingInitialSplits.get() > 0) {
-                    maxBytes = maxInitialSplitSize.toBytes();
-                    creatingInitialSplits = true;
-                }
+                @Override
+                protected HiveSplit computeNext()
+                {
+                    if (!blockLocationIterator.hasNext()) {
+                        return endOfData();
+                    }
+                    BlockLocation blockLocation = blockLocationIterator.peek();
 
-                // divide the block into uniform chunks that are smaller than the max split size
-                int chunks = toIntExact((long) Math.ceil(blockLocation.getLength() * 1.0 / maxBytes));
-                // when block does not divide evenly into chunks, make the chunk size slightly bigger than necessary
-                long targetChunkSize = (long) Math.ceil(blockLocation.getLength() * 1.0 / chunks);
-
-                long chunkOffset = 0;
-                while (chunkOffset < blockLocation.getLength()) {
-                    if (remainingInitialSplits.decrementAndGet() < 0 && creatingInitialSplits) {
```

Comment:
You're right. I should add the decrementing back.

Do you have any suggestions on verifying correctness of this code. I tried adding tests yesterday. I guess I can sit down and take time to manually construct a BackgroundHiveSplitLoader if that's what we have to do. Do you have any suggestions?